backend/src/common/services/calculation-progress/calculation-progress.service.ts

```diff
@@ -0,0 +1,154 @@
+    private goalService: GoalService;
+    private userAchievementsService: UserAchievementService;
+
+    public constructor(
+        goalService: GoalService,
+        achievementService: AchievementService,
+        userAchievementsService: UserAchievementService,
+    ) {
+        this.achievementService = achievementService;
+        this.goalService = goalService;
+        this.userAchievementsService = userAchievementsService;
+    }
+
+    public async calculateProgress(userId: number): Promise<void> {
+        const { items: goals } = await this.goalService.findAll({ userId });
+        const { items: workouts } = await workoutService.findAll({ userId });
+
+        await this.updateGoal(userId, goals, workouts);
+        await this.updateAchievement(userId, workouts);
+    }
+
+    private async updateGoal(
+        userId: number,
+        goals: GoalResponseDto[],
+        workouts: WorkoutResponseDto[],
+    ): Promise<void> {
+        for (const goal of goals) {
+            const filteredWorkouts = workouts.filter(
+                (workout) => workout.activityType === goal.activityType,
+            );
+
+            await this.goalService.update(
+                { id: goal.id },
+                {
+                    ...goal,
+                    userId,
+                    progress: calculateGoalProgress(goal, filteredWorkouts),
+                    completedAt:
+                        calculateGoalProgress(goal, filteredWorkouts) ===
+                        COMPLETED_GOAL_VALUE
+                            ? new Date().toISOString()
+                            : null,
+                },
+            );
+        }
+    }
+
+    private async updateAchievement(
+        userId: number,
+        workouts: WorkoutResponseDto[],
+    ): Promise<void> {
+        const achievementsList = await this.achievementService.findAll();
+        const userAchievementsList =
+            await this.achievementService.findByUserId(userId);
+        const userAchievementsListById = userAchievementsList?.map(
+            (item) => item.toObject().id,
+        );
+
+        const userAchievements = [];
+
+        const cyclingWorkouts = filterWorkoutsByActivityType(
+            workouts,
+            ActivityType.CYCLING,
+        );
+        const cyclingAchievements = filterAchievementByActivityType(
+            achievementsList,
+            ActivityType.CYCLING,
+        );
+        const userCyclingAchievements = cyclingAchievements
+            .filter((achievement) =>
+                checkCyclingAchievements(cyclingWorkouts, achievement),
+            )
+            .map((item) => item.toObject().id);
+        const uniqueCyclingAchievements = checkAchievementUniqueness(
+            userCyclingAchievements,
+            userAchievementsListById,
+        );
+        if (uniqueCyclingAchievements.length > 0) {
+            userAchievements.push(...uniqueCyclingAchievements);
+        }
```

Comment:
You can simplify it like this:  
`userAchievements.push(...uniqueCyclingAchievements);`
Spreading an empty array into another array has no effect, so the condition is not necessary.